File: apps/mobile/lib/src/sync/sync_coordinator.dart
import 'contracts.dart';
import 'local_sync_repository.dart';

/// Transport boundary for the frozen v1 sync protocol. An HTTP, gRPC, or
/// platform-specific client can implement this without leaking its response
/// types into the repository or domain code.
abstract interface class SyncTransport {
  Future<SyncChangePage> pull(
    SyncPartition partition,
    int cursor, {
    int limit = 500,
  });

  Future<SyncCommandReceipt> submit(
    SyncPartition partition,
    OutboxCommand command,
  );
}

final class SyncRunResult {
  const SyncRunResult({
    required this.pagesPulled,
    required this.changesApplied,
    required this.commandsSubmitted,
    required this.commandsApplied,
    required this.commandsRejected,
    required this.commandsConflicted,
    required this.commandsRetryPending,
  });

  final int pagesPulled;
  final int changesApplied;
  final int commandsSubmitted;
  final int commandsApplied;
  final int commandsRejected;
  final int commandsConflicted;
  final int commandsRetryPending;
}

/// Runs one bounded, idempotent reconnect pass.
///
/// Storage owns durability and transport owns authentication/retry policy. If
/// a network call fails after a command is marked sending, the command is
/// recovered to pending with the same id before the error is rethrown.
final class SyncCoordinator {
  SyncCoordinator(
    this._repository,
    this._transport, {
    int pageSize = 500,
    int commandBatchSize = 50,
    int maxPages = 1000,
  })  : _pageSize = _bound(pageSize, 500, 'pageSize'),
        _commandBatchSize = _bound(commandBatchSize, 500, 'commandBatchSize'),
        _maxPages = _bound(maxPages, 10000, 'maxPages');

  final LocalSyncRepository _repository;
  final SyncTransport _transport;
  final int _pageSize;
  final int _commandBatchSize;
  final int _maxPages;

  Future<SyncRunResult> reconnect(SyncPartition partition) async {
    var pagesPulled = 0;
    var changesApplied = 0;
    var commandsSubmitted = 0;
    var commandsApplied = 0;
    var commandsRejected = 0;
    var commandsConflicted = 0;
    var commandsRetryPending = 0;

    await _repository.recoverInterrupted(partition);
    final first = await _pullUntilCurrent(partition);
    pagesPulled += first.pages;
    changesApplied += first.changes;

    try {
      while (true) {
        final pending = await _repository.pending(partition, limit: _commandBatchSize);
        if (pending.isEmpty) break;
        for (final command in pending) {
          await _repository.markSending(partition, command.commandId);
          final receipt = await _transport.submit(partition, command);
          commandsSubmitted += 1;
          if (receipt.status == SyncReceiptStatus.retryPending) {
            commandsRetryPending += 1;
            await _repository.recoverInterrupted(partition);
            return SyncRunResult(
              pagesPulled: pagesPulled,
              changesApplied: changesApplied,
              commandsSubmitted: commandsSubmitted,
              commandsApplied: commandsApplied,
              commandsRejected: commandsRejected,
              commandsConflicted: commandsConflicted,
              commandsRetryPending: commandsRetryPending,
            );
          }
          final status = switch (receipt.status) {
            SyncReceiptStatus.applied => OutboxStatus.applied,
            SyncReceiptStatus.rejected => OutboxStatus.rejected,
            SyncReceiptStatus.conflict => OutboxStatus.conflict,
            SyncReceiptStatus.retryPending => OutboxStatus.retryPending,
          };
          await _repository.resolve(partition, command.commandId, status, errorCode: receipt.resultCode);
          if (receipt.status == SyncReceiptStatus.applied) commandsApplied += 1;
          if (receipt.status == SyncReceiptStatus.rejected) commandsRejected += 1;
          if (receipt.status == SyncReceiptStatus.conflict) commandsConflicted += 1;
        }
      }
    } catch (_) {
      await _repository.recoverInterrupted(partition);
      rethrow;
    }

    final finalPull = await _pullUntilCurrent(partition);
    pagesPulled += finalPull.pages;
    changesApplied += finalPull.changes;
    return SyncRunResult(
      pagesPulled: pagesPulled,
      changesApplied: changesApplied,
      commandsSubmitted: commandsSubmitted,
      commandsApplied: commandsApplied,
      commandsRejected: commandsRejected,
      commandsConflicted: commandsConflicted,
      commandsRetryPending: commandsRetryPending,
    );
  }

  Future<({int pages, int changes})> _pullUntilCurrent(SyncPartition partition) async {
    var cursor = await _repository.cursor(partition);
    var pages = 0;
    var changes = 0;
    while (true) {
      pages += 1;
      if (pages > _maxPages) throw const SyncRepositoryException('sync_page_limit_exceeded');
      final page = await _transport.pull(partition, cursor, limit: _pageSize);
      if (page.version != syncProtocolVersion) throw const SyncRepositoryException('unsupported_protocol_version');
      if (page.cursor < cursor) throw const SyncRepositoryException('regressing_sync_cursor');
      if (page.changes.isNotEmpty && page.changes.last.cursor > page.cursor) {
        throw const SyncRepositoryException('change_cursor_exceeds_page_cursor');
      }
      final before = cursor;
      await _repository.applyChanges(partition, cursor, page.changes);
      changes += page.changes.length;
      cursor = await _repository.cursor(partition);
      if (!page.hasMore) return (pages: pages, changes: changes);
      if (page.changes.isEmpty || cursor <= before) {
        throw const SyncRepositoryException('sync_page_did_not_advance');
      }
    }
  }
}

int _bound(int value, int maximum, String name) {
  if (value < 1 || value > maximum) {
    throw ArgumentError.value(value, name, 'must be between 1 and $maximum');
  }
  return value;
}

File: apps/web/src/offline/sync-coordinator.ts
import type { SyncRepository } from "./sync-repository.ts";
import type {
  ChangeBatch,
  OutboxEntry,
  SyncCommandEnvelope,
  SyncCommandReceipt,
  SyncPartition,
} from "./sync-contract.ts";

export type SyncTransport = Readonly<{
  pull: (
    partition: SyncPartition,
    cursor: number,
    limit?: number,
  ) => Promise<ChangeBatch>;
  submit: (
    partition: SyncPartition,
    command: SyncCommandEnvelope,
  ) => Promise<SyncCommandReceipt>;
}>;

export type SyncRunResult = Readonly<{
  pagesPulled: number;
  changesApplied: number;
  commandsSubmitted: number;
  commandsApplied: number;
  commandsRejected: number;
  commandsConflicted: number;
  commandsRetryPending: number;
}>;

export type SyncCoordinatorOptions = Readonly<{
  pageSize?: number;
  commandBatchSize?: number;
  maxPages?: number;
  now?: () => Date;
}>;

const positiveBound = (value: number, maximum: number, name: string): number => {
  if (!Number.isSafeInteger(value) || value < 1 || value > maximum) {
    throw new Error(`${name} must be between 1 and ${maximum}`);
  }
  return value;
};

/**
 * Coordinates one reconnect pass without owning transport, storage, or domain
 * policy. It is deliberately safe to call again after a network interruption:
 * outbox command ids and server cursors make the second pass idempotent.
 */
export class SyncCoordinator {
  readonly #repository: SyncRepository;
  readonly #transport: SyncTransport;
  readonly #pageSize: number;
  readonly #commandBatchSize: number;
  readonly #maxPages: number;
  readonly #now: () => Date;

  constructor(
    repository: SyncRepository,
    transport: SyncTransport,
    options: SyncCoordinatorOptions = {},
  ) {
    this.#repository = repository;
    this.#transport = transport;
    this.#pageSize = positiveBound(options.pageSize ?? 500, 500, "pageSize");
    this.#commandBatchSize = positiveBound(
      options.commandBatchSize ?? 50,
      500,
      "commandBatchSize",
    );
    this.#maxPages = positiveBound(options.maxPages ?? 1000, 10_000, "maxPages");
    this.#now = options.now ?? (() => new Date());
  }

  async reconnect(partition: SyncPartition): Promise<SyncRunResult> {
    const result = {
      pagesPulled: 0,
      changesApplied: 0,
      commandsSubmitted: 0,
      commandsApplied: 0,
      commandsRejected: 0,
      commandsConflicted: 0,
      commandsRetryPending: 0,
    };

    // A tab or process can disappear after claiming work. Recover it before
    // reading the queue so reconnect never strands a command in `sending`.
    await this.#repository.recoverInterrupted(partition, this.#now());
    await this.#pullUntilCurrent(partition, result);

    try {
      await this.#flushOutbox(partition, result);
    } catch (error) {
      // The current command and any commands claimed beside it are safe to
      // retry with the same ids after a network/transport interruption.
      await this.#repository.recoverInterrupted(partition, this.#now());
      throw error;
    }

    // A successful command can publish a change (for example a finalized
    // sale), so finish with another pull before reporting the pass complete.
    await this.#pullUntilCurrent(partition, result);
    return result;
  }

  async #pullUntilCurrent(
    partition: SyncPartition,
    result: { pagesPulled: number; changesApplied: number },
  ): Promise<void> {
    let cursor = await this.#repository.cursor(partition);
    let pages = 0;
    while (true) {
      pages += 1;
      if (pages > this.#maxPages) {
        throw new Error("Sync pull exceeded its safety page limit");
      }
      const page = await this.#transport.pull(partition, cursor, this.#pageSize);
      if (page.cursor < cursor) {
        throw new Error("Sync pull returned a regressing cursor");
      }
      await this.#repository.applyChanges(partition, page);
      result.pagesPulled += 1;
      result.changesApplied += page.changes.length;
      cursor = page.cursor;
      if (!page.hasMore) return;
      if (page.cursor === (await this.#repository.cursor(partition)) && page.changes.length === 0) {
        throw new Error("Sync pull reported more pages without advancing");
      }
    }
  }

  async #flushOutbox(
    partition: SyncPartition,
    result: {
      commandsSubmitted: number;
      commandsApplied: number;
      commandsRejected: number;
      commandsConflicted: number;
      commandsRetryPending: number;
    },
  ): Promise<void> {
    while (true) {
      const entries = await this.#repository.claimPending(
        partition,
        this.#commandBatchSize,
        this.#now(),
      );
      if (entries.length === 0) return;
      for (const entry of entries) {
        const deferred = await this.#submitEntry(partition, entry, result);
        // Do not spin on a server-deferred command. It is pending again and
        // belongs to the next reconnect/backoff window.
        if (deferred) return;
      }
    }
  }

  async #submitEntry(
    partition: SyncPartition,
    entry: OutboxEntry,
    result: {
      commandsSubmitted: number;
      commandsApplied: number;
      commandsRejected: number;
      commandsConflicted: number;
      commandsRetryPending: number;
    },
  ): Promise<boolean> {
    const receipt = await this.#transport.submit(partition, entry.envelope);
    result.commandsSubmitted += 1;
    if (receipt.status === "retry_pending") {
      result.commandsRetryPending += 1;
      await this.#repository.recoverInterrupted(partition, this.#now());
      return true;
    }
    await this.#repository.settle(
      partition,
      entry.envelope.commandId,
      receipt.status,
      receipt.resultCode,
    );
    if (receipt.status === "applied") result.commandsApplied += 1;
    if (receipt.status === "rejected") result.commandsRejected += 1;
    if (receipt.status === "conflict") result.commandsConflicted += 1;
    return false;
  }
}
